### DISCUSSION TITLE
Looots of files in "doublecmd\pixmaps" folder (1089 files) -- Is it possible to put one DLL file (or uncompressed ZIP)?

### DISCUSSION THREAD

**Discussion by 999px** [doublecmd/doublecmd#1845] [General]
### Version

Last Windows version

### Description

Looots of files in "doublecmd\pixmaps" folder (1089 files).
Is it possible to put  one DLL file?

**Comment by j2969719** [comment]
doublecmd is not a windows-only application. the icon theme structure used here is similar to linux icon themes and works under any suported platform, unlike dll

**Reply by 999px** [reply]
I didn't think about that.
But I agree with you, in such a case DLL is a bad idea.
But maybe some other format can be used in such a case.
For example - uncompressed ZIP.

**Comment by ags1234** [comment]
Those pixmaps could probably be stored in a SQLITE database as BLOBs.
But being like they are now on the disk, separate files, is more open and easier to change something in there. One does not have to deal with a database overhead.
Though I can understand OP's mindset, I don't see a real benefit in this change, so I would be in favor of letting the files as they are now.

**Reply by 999px** [reply]
Maybe then make it even simpler
so that icons are read both from a folder (as now) and from a ZIP archive.
that is, icons can be either here: \doublecmd\pixmaps\
or here: \doublecmd\pixmaps.zip

**Comment by alexx2000** [comment]
Database is the worst variant. Zip-archive is optimal, one file and easy to modify.

**Reply by 999px** [reply]
Yeah, I agree with you.

**Comment by ags1234** [comment]
I wonder what would be the cost in milliseconds delayed with zip archives?

**Reply by alexx2000** [reply]
I assume that it can be faster than a file system in general.

**Comment by ags1234** [comment]
Faster than a file system? How can this be possible?
With zip archives there are the extra steps of loading the zip in RAM and decompressing the content there. This takes time that is not consumed when the files are accessed directly.
Or do you mean to say that accessing the files on disk is slower than loading, decompressing and accessing in RAM?

**Reply by JonnyTech** [reply]
Disk operations are expensive, especially when multiplied by over 1000. RAM will always beat that.

I would like to see this implemented, not specifically to reduce file count, but to improve installation times. Copying all those files to a slow USB memory drive or via a slow FTP connection is very timesome because of the multiple IO operations. A single larger file would benefit me immensly.

**Reply by 999px** [reply]
[JonnyTech](https://github.com/JonnyTech)
Maybe then this can be implemented in the new version?
Where is the best place to write so that the developer can see this request?
or is it enough to just click on “Vote Up” and they will see?

**Reply by alexx2000** [reply]
> the developer can see this request?

It's me :)

**Comment by JonnyTech** [comment]
@alexx2000 any plans on the roadmap to implement a single zip file? I suspect that it may aid issues such as #2297

**Reply by alexx2000** [reply]
In case of #2297 I guess it does not help much. I think that the main problem is `ePapirus` icon theme, which are very large (about 5000 icons). Some kind of caching mechanism needs to be implemented.

**Reply by JonnyTech** [reply]
Understood, thanks, although what about my use case in https://github.com/doublecmd/doublecmd/discussions/1845#discussioncomment-10711453 is there an alternative apart from appimage?

**Reply by alexx2000** [reply]
I guess - no.

**Comment by 999px** [comment]
Could you perhaps make it possible to replace the doublecmd\pixmaps folder with a single (uncompressed) ZIP archive?

My perfectionism really dislikes seeing more than 1000 files.

**Reply by alexx2000** [reply]
I will try to do it.

**Reply by 999px** [reply]
> I will try to do it.

Thank you very much! :)

**Comment by eMouses** [comment]
I'll add one argument in favor of a single file — a large number of small files significantly slows down copying to an external storage device (such as a USB flash drive).

**Reply by ags1234** [reply]
I was added already, see what @JonnyTech said [above](https://github.com/doublecmd/doublecmd/discussions/1845#discussioncomment-10711453).
:))

I am also glad this is being investigated and worked on.
There is a lot of good work being done to DC lately. Thank you for that.

**Comment by Skif-off** [comment]
> Copying all those files to a slow USB memory drive or via a slow FTP connection is very timesome

> I'll add one argument in favor of a single file — a large number of small files significantly slows down copying to an external storage device (such as a USB flash drive).


How many times a day do you have to copy Double Commander to a [slow] USB flash drive? :) And why?

I use x64 GTK2 version and I did a little research by downloading several versions, so let's consider updating DC on a USB flash drive:
```
Ver.                 All    Diff    Add    Rem
1.0.0  (2021.10.17)  1202
1.0.10 (2023.01.21)  1201    50
1.1.0  (2023.08.05)  1257    287    +62
1.1.10 (2024.02.17)  1240    78     +5     -22
1.1.20 (2024.11.10)  1244    29     +5     -1
1.1.31 (2025.12.21)  1295    26     +52    -1
1.1.32 (2026.01.07)  1295    5
1.2.0  (2026.01.11)  1308    71     +13
1.2.1  (2026.01.25)  1308    6
1.2.2  (2026.02.15)  1308    3
1.2.3  (2026.03.01)  1327    5      +19
1.2.4  (2026.03.14)  1327    6
1.2.6  (2026.05.10)  1345    43     +18
1.3.0  (2026.07.19)  1364    83     +1
```
---

I don't see a _very big_ problem, but I like the idea of ​​packaging `pixmaps` in a ZIP archive (like the standard Python libraries, for example).

@alexx2000, if this will added, will the current behavior be preserved? Will replacing icons with system theme icons also be preserved?
What about third-party icons theme?

**Reply by alexx2000** [reply]
Mostly yes. DC icon theme will have an optional `zip` support (capability to store icons directories `8x8 - scalable` in the archive). I already implement this feature (draft). I will publish it when it will be completely ready.

**Reply by eMouses** [reply]
> How many times a day do you have to copy Double Commander to a [slow] USB flash drive? :) And why?
> 
> I don't see a _very big_ problem, but I like the idea of ​​packaging `pixmaps` in a ZIP archive (like the standard Python libraries, for example).
> 
I periodically copy the entire DC folder to another machine — the whole folder at once (not by incremental update). The issue may not be huge, but a single file is definitely more convenient either way.

**Comment by alexx2000** [comment]
I added an optional `zip` support for DCTheme. How to test:
1. Go to `pixmaps/dctheme` directory;
2. Pack `8x8 - scalable` directories to `icon-theme.zip` archive without compression;
3. Remove `8x8 - scalable` directories and `icon-theme.cache` file;
4. Restart Double Commander.

**Reply by eMouses** [reply]
What is the file icon-theme.cache for?

**Reply by alexx2000** [reply]
For fast loading icons list. In case of zip-archive it is not needed, because zip-archive also works as cache.

**Reply by Skif-off** [reply]
@alexx2000, can you try to run DC with custom ison theme? For example, ![dcthemeext.zip](https://github.com/user-attachments/files/30751150/dcthemeext.zip)

**Reply by alexx2000** [reply]
In what combination? Both compressed (default and custom), one compressed and one not?

**Reply by Skif-off** [reply]
I just updated the app (`doublecmd.exe`, `doublecmd.zdli` + `zip.wcx`) and couldn't get it to work until I changed the theme name in `doublecmd.xml` to the default value.

**Reply by alexx2000** [reply]
Fixed.

**Comment by ags1234** [comment]
> 2. Pack 8x8 - scalable directories to icon-theme.zip archive without compression;

How do I do this with DC itself?
There is only the option for Fastest compression (1), no option for no compression.

<img width="406" height="307" alt="2026-08-05_151023" src="https://github.com/user-attachments/assets/4ac99082-4ec9-4b4a-9015-985526845c10" />

**Reply by ags1234** [reply]
To respond my own question, I guess changing the compression method to Store should do it.

**Reply by alexx2000** [reply]
Yes

**Comment by eMouses** [comment]
Can custom icons be added to dctheme/icon-theme.zip? For example, for building menus and Lua scripts.

**Reply by alexx2000** [reply]
Default theme modification is a bad idea. The right way is to [create a custom theme](https://doublecmd.github.io/doc/en/faq.html#theme) that contains only needed icons. Like **Skif-off** uploaded above. These icons can be used in the toolbar buttons and menus.

**Reply by eMouses** [reply]
Got it. But is there any way to use custom filenames for icons? Or only those that are in *action*?

**Reply by alexx2000** [reply]
You can use any file names for custom actions.

**Reply by eMouses** [reply]
When I specify the path to the icon as
%COMMANDER_PATH%\pixmaps\dcthemetest\16x16\actions\calculator.png
then it works.

But if I specify just calculator.png or calculator, DC does not see the icon. I have set the theme to dcthemetest in the settings.

**Reply by Skif-off** [reply]
Forgive my curiosity...
Default icon theme doesn't contain any icons with that name: why are you doing this?

**Reply by eMouses** [reply]
For example, there are Lua scripts for which I would like to have my own icons with meaningful names.

**Reply by alexx2000** [reply]
I guess you are doing something wrong. Works fine:

<img width="727" height="709" alt="изображение" src="https://github.com/user-attachments/assets/0ecacc0f-0718-4e8b-ae8b-f1603b2a9c51" />

**Reply by eMouses** [reply]
I'm restarted DC, and then it recognized the icon I needed. Thank you for the help!

**Comment by 999px** [comment]
First of all, I would like to thank @alexx2000 . Thank you very much for this update! :)

As I understand it, you don't want to replace the icon folders with a ZIP archive by default.

Then perhaps it would be a good idea to create a `THIS_FOLDER_INFO.txt` file in `<Double Commander>\pixmaps\` with the following content:

***********************
If you want to reduce the number of files in this folder, follow these steps:

1. Go to `<Double Commander>\pixmaps\dctheme\`

2. Delete the `icon-theme.cache` file

3. Select all the folders (`8x8`, `16x16` ... `scalable`) and add them to a ZIP-archive - "Store" (no compression).
   Name the archive `icon-theme.zip`

4. Delete the aforementioned folders (`8x8`, `16x16` ... `scalable`).

5. If everything was done correctly, the `<Double Commander>\pixmaps\dctheme\`
   folder will contain only two files: `icon-theme.zip` and `index.theme`.

That's it.
***********************************************

It would be a shame for all the work that went into this to go unnoticed, with only a few people here knowing about it.

**Reply by alexx2000** [reply]
I plan to enable it by default in the future. This is a new feature, so it should be tested first. I will enable it in the snapshots soon.
